src/common/buffer.cc

```diff
@@ -399,183 +458,380 @@ static simple_spinlock_t buffer_debug_lock = SIMPLE_SPINLOCK_INITIALIZER;
-      dec_total_alloc(len);
-      bdout << "raw_pipe " << this << " free " << (void *)data << " "
-	    << buffer::get_total_alloc() << bendl;
+  void append_pipe() {
+    int r;
+    if (pipe_curr_write_fd != -1) {
+      close(pipe_curr_write_fd);
+    }
+    int fds[2];
+    if (::pipe(fds) == -1) {
+      r = -errno;
+      bdout << "raw_pipe: error creating pipe: " << cpp_strerror(r) << bendl;
+      throw error_code(r);
+    }
+    r = ::fcntl(fds[write_end], F_SETFL, O_NONBLOCK);
+    if (r != -1)
+      r = ::fcntl(fds[read_end], F_SETFL, O_NONBLOCK);
+    if (r < 0) {
+      bdout << "raw_pipe: error setting nonblocking flag on temp pipe: "
+          << cpp_strerror(r) << bendl;
+      throw error_code(r);
     }
+    pipe_curr_write_fd = fds[write_end];
+    set_pipe_size(pipe_curr_write_fd);
+    pipe_read_elements.push_back({0, fds[read_end]});
+  }
 
-    bool can_zero_copy() const {
-      return true;
+
+  bool can_zero_copy() const {
+    return data == nullptr;
+  }
+
+  ssize_t consume_pipe(int fd, size_t len)
+  {
+    int devnull;
+    devnull = open("/dev/null", O_WRONLY);
+    if (devnull == -1)
+      return -1;
+    ssize_t r;
+    r = splice(fd, nullptr, devnull, nullptr, len, SPLICE_F_MOVE);
+    ::close(devnull);
+    return r;
+  }
+
+  bool is_pipe_full(int fd)
+  {
+    struct pollfd fds;
+    fds.fd = fd;
+    fds.events = POLLOUT;
+    fds.revents = 0;
+    int r = poll(&fds, 1, 0);
+    //when no fds.revents are set, means we are blocked
+    return (r == 0);
+  }
+
+  /*
+   * Streams data from buffer to specified file.
+   * If destination file is blocking, function will block until all data is transferred.
+   * If destination file is non-blocking, function will transfer as much data as possible.
+   *
+   * \note Transferred data is not lost from buffer.
+   *
+   * \param fd file to write to
+   * \param src_pos start streaming from this byte
+   * \param count amount of bytes to write
+   * \param dst_offset offset in \ref fd to write to. use -1 to append to current file position.
+   *
+   * \return amount of bytes transferred, or -1 if operation would block
+   *
+   * \throws end_of_buffer if request is outside buffer's range
+   */
+  ssize_t copy_to_fd(int fd, off_t ofs_from, ssize_t req_size, off_t dst_offset) {
+    if (ofs_from + req_size > len) {
+      throw end_of_buffer();
     }
+    if (data != nullptr) {
+      return buffer::raw::copy_to_fd(fd, ofs_from, req_size, dst_offset);
+    }
+    //move to proper pipe
+    size_t i = 0;
+    off_t ofs = ofs_from;
+    for (i = 0; i < pipe_read_elements.size() && ofs >= (off_t)pipe_read_elements[i].size; i++ ) {
+      ofs -= pipe_read_elements[i].size;
+    }
+    assert(i < pipe_read_elements.size());
+    assert(ofs < (off_t)pipe_read_elements[i].size);
+    ssize_t rem_size = req_size;
+    ssize_t total_moved = 0;
+    while (rem_size > 0) {
+      assert(i < pipe_read_elements.size());
+      int fds[2];
+      if (::pipe(fds) == -1) {
```

Comment:
Statement
assert(i < pipe_read_elements.size());
actually checks validity of check stated in beginning of function:
if (ofs_from + req_size > len) {
   throw end_of_buffer();
}
This is verifying internal data consistency.

Should I change it to throw end_of_buffer() when assert condition trigger?